presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java

```diff
@@ -0,0 +1,332 @@
+    {
+        presentStreamSource = missingStreamSource(BooleanInputStream.class);
+        dataStreamSource = missingStreamSource(BooleanInputStream.class);
+
+        readOffset = 0;
+
+        presentStream = null;
+        dataStream = null;
+
+        rowGroupOpen = false;
+    }
+
+    @Override
+    public void startRowGroup(InputStreamSources dataStreamSources)
+    {
+        presentStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, PRESENT, BooleanInputStream.class);
+        dataStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, DATA, BooleanInputStream.class);
+
+        readOffset = 0;
+
+        presentStream = null;
+        dataStream = null;
+
+        rowGroupOpen = false;
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .addValue(streamDescriptor)
+                .toString();
+    }
+
+    @Override
+    public void close()
+    {
+        systemMemoryContext.close();
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + sizeOf(values) + sizeOf(nulls) + sizeOf(outputPositions);
+    }
+
+    private void openRowGroup()
+            throws IOException
+    {
+        presentStream = presentStreamSource.openStream();
+        dataStream = dataStreamSource.openStream();
+        rowGroupOpen = true;
+    }
+
+    @Override
+    public int read(int offset, int[] positions, int positionCount)
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        boolean nullsAllowed = filter == null || filter.testNull();
+        boolean recordNulls = nullsAllowed && presentStream != null;
+        if (outputRequired) {
+            ensureValuesCapacity(positionCount, recordNulls);
+        }
+
+        if (filter != null) {
+            ensureOutputPositionsCapacity(positionCount);
+        }
+        else {
+            outputPositions = positions;
+        }
+
+        if (readOffset < offset) {
+            skip(offset - readOffset);
+        }
+
+        int streamPosition = 0;
+        allNulls = false;
+        if (dataStream == null && presentStream != null) {
+            // all nulls
+            presentStream.skip(positions[positionCount - 1]);
+            streamPosition = positions[positionCount - 1] + 1;
+
+            if (nullsAllowed) {
+                outputPositionCount = positionCount;
+                allNulls = true;
+            }
+            else {
+                outputPositionCount = 0;
+            }
+        }
+        else if (filter == null) {
+            for (int i = 0; i < positionCount; i++) {
+                int position = positions[i];
+                if (position > streamPosition) {
+                    skip(position - streamPosition);
+                    streamPosition = position;
+                }
+
+                if (presentStream != null && !presentStream.nextBit()) {
```

Comment:
The reason I was curious about https://github.com/prestodb/presto/pull/12991/commits/77178e4947d85560cecdc0e9ba092f4e1f1c2c3f#r298004032 is because there are a few if/elses in this logic. Some of them need some thinking before understanding what it does exactly. Possibly it's due to the different combination of `null X filter X outputRequired`. Can we extract some conditions and assignments into helper functions? For example, `if (presentStream != null && !presentStream.nextBit())` could be `private boolean nextPositionIsNull()`. Some of the if/else checks could be hidden in the helper functions for easy understanding.

Reply:
- `filter == null` implies `outputRequired == true`, because if there is no filter and no output is needed, then the reader should not have been created in the first place
- `nextPositionIsNull()` will have a side effect of advancing the presentStream; are we OK with that?
- I agree that the are a bunch of flags and combinations that need to be handled carefully

Reply:
Added `checkArgument(filter.isPresent() || outputRequired, "filter must be present if outputRequired is false");` to the constructor of the `BooleanSelectiveStreamReader`.